page.clearMemoryCache(): clearMemoryCache() + clearCache() + loadImages defineProperty

hlr1983 opened the pull request:
I think it might be useful, as Changelog says newPage() was created as "option to create a new WebPage instance to address memory consumption issues (#803, #826)".

jefleponot commented on modules/casper.js:
```diff
@@ -2513,6 +2513,7 @@ Casper.prototype.newPage = function newPage() {
     if (utils.isObject(this.options.viewportSize)) {
         this.page.viewportSize = this.options.viewportSize;
     }
+    this.page.clearMemoryCache()
```
uncaughtError: TypeError: 'undefined' is not a function (evaluating 'this.page.clearMemoryCache()')

hlr1983 replied:
It worked for me, running in Linux/Ubuntu/Xenial, with phantomjs 1.98, 2.1.1 and 2.5:

henrique@usb-azul:~/test_casper$ lsb_release -a ; PATH=$PATH:./phantomjs/ ; echo "casper.js with this.page.clearMemoryCache():" ; grep -B16 -A3 MemoryCache ./casperjs/modules/casper.js ; echo "My test script" ; cat ./test.js ; cp phantomjs/phantomjs-250 phantomjs/phantomjs ; phantomjs --version ; ./casperjs/bin/casperjs ./test.js ; cp phantomjs/phantomjs-211 phantomjs/phantomjs ; phantomjs --version ; ./casperjs/bin/casperjs ./test.js ; cp phantomjs/phantomjs-198 phantomjs/phantomjs ; phantomjs --version ; ./casperjs/bin/casperjs ./test.js 
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.2 LTS
Release:	16.04
Codename:	xenial
casper.js with this.page.clearMemoryCache():
* Allow user to create a new page object after calling a casper.page.close()
* @return WebPage
*/

Casper.prototype.newPage = function newPage() {
    "use strict";
    this.checkStarted();
    this.page.close();
    this.page = this.mainPage = createPage(this);
    this.page.settings = utils.mergeObjects(this.page.settings, this.options.pageSettings);
    if (utils.isClipRect(this.options.clipRect)) {
        this.page.clipRect = this.options.clipRect;
    }
    if (utils.isObject(this.options.viewportSize)) {
        this.page.viewportSize = this.options.viewportSize;
    }
    this.page.clearMemoryCache()
    return this.page;
};

My test script

var casper = require('casper').create({
	verbose: true,
	logLevel: "debug",
});

casper.start("");
casper.thenOpen("http://www.google.com/", function () { this.capture("test1.png"); });
casper.then( function () { this.page = this.newPage(); });
casper.thenOpen("http://www.yahoo.com", function () { this.capture("test2.png"); });
casper.run();
2.5.0-development
[info] [phantom] Starting...
[info] [phantom] Running suite: 5 steps
[debug] [phantom] opening url: http://www.google.com/, HTTP GET
[debug] [phantom] Navigation requested: url=http://www.google.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://www.google.com/"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 2/5 http://www.google.com/ (HTTP 0)
[debug] [phantom] Capturing page to /home/henrique/test_casper/test1.png
[info] [phantom] Capture saved to /home/henrique/test_casper/test1.png
[info] [phantom] Step anonymous 2/5: done in 3403ms.
[info] [phantom] Step anonymous 3/5 http://www.google.com/ (HTTP 0)
[info] [phantom] Step anonymous 3/5: done in 3421ms.
[debug] [phantom] opening url: http://www.yahoo.com/, HTTP GET
[debug] [phantom] Navigation requested: url=http://www.yahoo.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://www.yahoo.com/"
[debug] [phantom] Navigation requested: url=https://s.yimg.com/os/mit/media/m/base/images/transparent-95031.png, type=Other, willNavigate=true, isMainFrame=false
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 5/5 http://www.yahoo.com/ (HTTP 0)
[debug] [phantom] Capturing page to /home/henrique/test_casper/test2.png
[info] [phantom] Capture saved to /home/henrique/test_casper/test2.png
[info] [phantom] Step anonymous 5/5: done in 7298ms.
[info] [phantom] Done 5 steps in 7366ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
2.1.1
[info] [phantom] Starting...
[info] [phantom] Running suite: 5 steps
[debug] [phantom] opening url: http://www.google.com/, HTTP GET
[debug] [phantom] Navigation requested: url=http://www.google.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] Navigation requested: url=http://www.google.com.br/?gws_rd=cr&ei=veHHWO3fEMa-wATeyLTgAw, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://www.google.com.br/?gws_rd=cr&ei=veHHWO3fEMa-wATeyLTgAw"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 2/5 http://www.google.com.br/?gws_rd=cr&ei=veHHWO3fEMa-wATeyLTgAw (HTTP 200)
[debug] [phantom] Capturing page to /home/henrique/test_casper/test1.png
[info] [phantom] Capture saved to /home/henrique/test_casper/test1.png
[info] [phantom] Step anonymous 2/5: done in 1622ms.
[info] [phantom] Step anonymous 3/5 http://www.google.com.br/?gws_rd=cr&ei=veHHWO3fEMa-wATeyLTgAw (HTTP 200)
[info] [phantom] Step anonymous 3/5: done in 1628ms.
[debug] [phantom] opening url: http://www.yahoo.com/, HTTP GET
[debug] [phantom] Navigation requested: url=http://www.yahoo.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] Navigation requested: url=https://www.yahoo.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] Navigation requested: url=https://br.yahoo.com/?p=us, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "https://br.yahoo.com/?p=us"
[debug] [phantom] Navigation requested: url=https://s.yimg.com/os/mit/media/m/base/images/transparent-95031.png, type=Other, willNavigate=true, isMainFrame=false
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 5/5 https://br.yahoo.com/?p=us (HTTP 200)
[debug] [phantom] Capturing page to /home/henrique/test_casper/test2.png
[info] [phantom] Capture saved to /home/henrique/test_casper/test2.png
[info] [phantom] Step anonymous 5/5: done in 4424ms.
[info] [phantom] Done 5 steps in 4485ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
1.9.8
[info] [phantom] Starting...
[info] [phantom] Running suite: 5 steps
[debug] [phantom] opening url: http://www.google.com/, HTTP GET
[debug] [phantom] Navigation requested: url=http://www.google.com/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] Navigation requested: url=http://www.google.com.br/?gws_rd=cr&ei=wuHHWJDIAoajwAT-toTYBg, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://www.google.com.br/?gws_rd=cr&ei=wuHHWJDIAoajwAT-toTYBg"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 2/5 http://www.google.com.br/?gws_rd=cr&ei=wuHHWJDIAoajwAT-toTYBg (HTTP 200)
[debug] [phantom] Capturing page to /home/henrique/test_casper/test1.png
[info] [phantom] Capture saved to /home/henrique/test_casper/test1.png
[info] [phantom] Step anonymous 2/5: done in 1519ms.
[info] [phantom] Step anonymous 3/5 http://www.google.com.br/?gws_rd=cr&ei=wuHHWJDIAoajwAT-toTYBg (HTTP 200)
Unsafe JavaScript attempt to access frame with URL about:blank from frame with URL file:///home/henrique/test_casper/casperjs/bin/bootstrap.js. Domains, protocols and ports must match.

jefleponot commented:
@hlr1983 
Have you find a documentation for that or just do reverse engineering ?

hlr1983 commented:
I've read about it in phantomjs changelog (https://github.com/ariya/phantomjs/blob/master/ChangeLog):

2015-01-23: Version 2.0.0

    New features

      * Switched to Qt 5 and updated WebKit (issue 10448)
      * Implemented clearing of memory cache (issue 10357)*
      * Added support for HTTP header change for every request (issue 11299)

And issue 10357 (https://github.com/ariya/phantomjs/issues/10357) says

"ariya added a commit that referenced this issue on Sep 25, 2014
 @ariya	Implement clearMemoryCache() on a web page to clear the cache.  …"

jefleponot commented:
@hlr1983 
You are right, good job.

An issue, why do you want to clean Memory for all new pages ?

jefleponot commented:
@hlr1983 
I was looking for the existence of the _newPage_ method. In fact, it was created when page.close event is called so navigation become impossible.
Clearing Cache could be really useful but it will be better to create a new casper's method : 

Casper.clearCache () , do you agree ?

hlr1983 commented:
Yes, I agree with creating the Casper.clearCache ().


I have not used this for address memory consumption issues.
I use casperjs for doing automation of tasks, and I had two situations where I had the behavior I wanted after adding a clearMemoryCache():

The following code, which only worked after adding a clearMemoryCache():
...
this.page.clearMemoryCache()
this.waitFor(
	function teste6() {
		return ( (this.exists(x("id('selSerie')/option[text()='Instrução'][@selected='selected']"))) && (this.exists(x("id('txtNumero')"))) && (this.exists(x("id('txtDataElaboracao')"))) && (this.exists(x("id('optPublico')"))) && (this.exists(x("id('btnSalvar')"))) )
	}, function sucesso6() {
		console.log("PREENCHENDO RESTANTE DO FORMULARIO E SALVANDO.");
		this.sendKeys(x("id('txtNumero')"), 'PROCESSUAL CADA', {keepFocus: true});
		this.sendKeys(x("id('txtDataElaboracao')"), data, {keepFocus: true});
		this.click(x("id('optPublico')"));
		this.click(x("id('btnSalvar')"));
		this.waitFor(
			function teste7() {
				if ( (this.exists(x("//a/span[@class='infraArvoreNoSelecionado']"))) && (this.getElementsInfo(x("//a/span[@class='infraArvoreNoSelecionado']")).length == 1) ) {
					return (/Instrução PROCESSUAL CADA \(\d{1,10}\)/.test(this.getElementInfo(x("//a/span[@class='infraArvoreNoSelecionado']")).text));
				} else {
					return false;
				};

			}, function sucesso7() {
...

the html code for XPath "id('btnSalvar')" is 

<button tabindex="451" type="button" onclick="confirmarDados()" accesskey="C" name="btnSalvar" id="btnSalvar" value="Confirmar Dados" class="infraButton"><span class="infraTeclaAtalho">C</span>onfirmar Dados</button>

I think it could be a misbehavior of confirmarDados() in phantomJS, because it works well when in chrome and firefox.



The other situation was when rendering a page that has the following html code:

<table cellspacing="0" cellpadding="0" height="23" border="1" width="599">
	<tbody><tr> 
		<td class="tableTitle1" align="center" width="599">Resultado da Consulta</td>
	</tr>
</tbody></table>

TableTitle1 is defined in a CSS file as

.tableTitle1
{
	border:1px solid #6D6C6C; background-color: #FFFFFF;
	font-family: Verdana, Arial, Helvetica, sans-serif;
	font-size: 10px;    
	font-weight : bold;	
	color: #ffffff;   
	text-decoration: none;
	text-transform: uppercase;
	height : 18px;
	padding-left : 0px;
	padding-top : 1px;	
	background-image: url('../images/frmshadow_txtfield1-on.jpg');
}

If I do something like

casper.then(function () { this.page.settings.loadImages=false; )};
casper.then(function () { this.open('A PAGE THAT LOADS THE CSS FILE AND USES TableTitle1'); )};
casper.then(function () { this.page.settings.loadImages=true; )};
casper.then(function () { this.open('THE PAGE I WANT TO RENDER THAT USES THE CSS TableTitle1'); )};
casper.then(function () { this.capture('page.png'); )};

The "background-image: url('../images/frmshadow_txtfield1-on.jpg');" is not shown in page.png

But if I do something like

casper.then(function () { this.page.settings.loadImages=false; )};
casper.then(function () { this.open('A PAGE THAT LOADS THE CSS FILE AND USES TableTitle1'); )};
casper.then(function () { this.page.clearMemoryCache(); )};
casper.then(function () { this.page.settings.loadImages=true; )};
casper.then(function () { this.open('THE PAGE I WANT TO RENDER THAT USES THE CSS TableTitle1'); )};
casper.then(function () { this.capture('page.png'); )};

The "background-image: url('../images/frmshadow_txtfield1-on.jpg');" is shown in page.png



I'm using phantomJS 2.1.1 in Linux x64, and the CasperJS 1.1.4

jefleponot commented:
Very good feedback....I will test your 2 cases ... Thanks a lot

hlr1983 commented:
I did a small script with offline files which reproduce the second case.
[test.tar.gz](https://github.com/casperjs/casperjs/files/849165/test.tar.gz)


```
henrique@usb-azul:~/test2$ rm *.png ; echo $PATH ; echo "MY TEST SCRIPT:" ; cat ./test.js ; ./phantomjs/phantomjs --version ; PATH=./phantomjs:$PATH ./casperjs/bin/casperjs ./test.js
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
MY TEST SCRIPT:

var casper = require("casper").create({
	verbose: true,
	logLevel: "debug",
});

var fs = require("fs");

casper.start("");

casper.then(function () { this.page.settings.loadImages=false; });
casper.then(function () { this.open("file:///" + fs.workingDirectory + "/www.site.com/multa/teste.html"); });
casper.then(function () { this.capture("page1.png"); });

casper.then(function () { this.page.settings.loadImages=true; });
casper.then(function () { this.open("file:///" + fs.workingDirectory + "/www.site.com/multa/teste.html"); });
casper.then(function () { this.capture("page2.png"); });

casper.then(function () { this.open("file:///" + fs.workingDirectory + "/www.site.com/multa/teste.html"); });
casper.then(function () { this.capture("page3.png"); });

casper.then(function () { this.page.clearMemoryCache() });
casper.then(function () { this.open("file:///" + fs.workingDirectory + "/www.site.com/multa/teste.html"); });
casper.then(function () { this.capture("page4.png"); });
casper.run();


2.1.1
[info] [phantom] Starting...
[info] [phantom] Running suite: 11 steps
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 1/11: done in 27ms.
[debug] [phantom] opening url: file:////home/henrique/test2/www.site.com/multa/teste.html, HTTP GET
[debug] [phantom] Navigation requested: url=file:////home/henrique/test2/www.site.com/multa/teste.html, type=Other, willNavigate=true, isMainFrame=true
[info] [phantom] Step anonymous 2/11: done in 45ms.
[debug] [phantom] url changed to "file:////home/henrique/test2/www.site.com/multa/teste.html"
[debug] [phantom] Successfully injected Casper client-side utilities
[debug] [phantom] Capturing page to /home/henrique/test2/page1.png
[info] [phantom] Capture saved to /home/henrique/test2/page1.png
[info] [phantom] Step anonymous 3/11: done in 125ms.
[info] [phantom] Step anonymous 4/11: done in 125ms.
[debug] [phantom] opening url: file:////home/henrique/test2/www.site.com/multa/teste.html, HTTP GET
[debug] [phantom] Navigation requested: url=file:////home/henrique/test2/www.site.com/multa/teste.html, type=Other, willNavigate=true, isMainFrame=true
[info] [phantom] Step anonymous 5/11: done in 146ms.
[debug] [phantom] url changed to "file:////home/henrique/test2/www.site.com/multa/teste.html"
[debug] [phantom] Successfully injected Casper client-side utilities
[debug] [phantom] Capturing page to /home/henrique/test2/page2.png
[info] [phantom] Capture saved to /home/henrique/test2/page2.png
[info] [phantom] Step anonymous 6/11: done in 192ms.
[debug] [phantom] opening url: file:////home/henrique/test2/www.site.com/multa/teste.html, HTTP GET
[debug] [phantom] Navigation requested: url=file:////home/henrique/test2/www.site.com/multa/teste.html, type=Other, willNavigate=true, isMainFrame=true
[info] [phantom] Step anonymous 7/11: done in 193ms.
[debug] [phantom] url changed to "file:////home/henrique/test2/www.site.com/multa/teste.html"
[debug] [phantom] Successfully injected Casper client-side utilities
[debug] [phantom] Capturing page to /home/henrique/test2/page3.png
[info] [phantom] Capture saved to /home/henrique/test2/page3.png
[info] [phantom] Step anonymous 8/11: done in 238ms.
[info] [phantom] Step anonymous 9/11: done in 240ms.
[debug] [phantom] opening url: file:////home/henrique/test2/www.site.com/multa/teste.html, HTTP GET
[debug] [phantom] Navigation requested: url=file:////home/henrique/test2/www.site.com/multa/teste.html, type=Other, willNavigate=true, isMainFrame=true
[info] [phantom] Step anonymous 10/11: done in 260ms.
[debug] [phantom] url changed to "file:////home/henrique/test2/www.site.com/multa/teste.html"
[debug] [phantom] Successfully injected Casper client-side utilities
[debug] [phantom] Capturing page to /home/henrique/test2/page4.png
[info] [phantom] Capture saved to /home/henrique/test2/page4.png
[info] [phantom] Step anonymous 11/11: done in 310ms.
[info] [phantom] Done 11 steps in 310ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
henrique@usb-azul:~/test2$ 

```

jefleponot commented:
@hlr1983 

Could ypu please do this test for me : 

add this line in the casper/tests/selftest.js line 54
```js
response.setEncoding("binary");
```

test the file casper/tests/suite/casper/loadImage.js
```js
/*eslint strict:0*/
casper.test.begin("load Images tests", 4, function(test) {
    casper.on("resource.received", function(resource) {
        if (resource.stage === "start") return;
        //console.log(JSON.stringify(resource,null,4));
    });

    casper.start().then(function(){
        casper.page.settings.loadImages = false;
        casper.thenOpen("tests/site/index.html", {},function() {
        test.assertEquals(this.resources.length, 1, "one resource found");
            this.wait(200, function() {
                this.capture('page0.png');
            });
        })
    }).then(function(){
        casper.page.settings.loadImages = true;
        casper.thenOpen("tests/site/index.html", {}, function() {
            test.assertNotEquals(this.resources.length, 1, "not only one found");
            this.wait(200, function() {
                this.capture('page1.png');
            });
        });
    }).then(function(){
        casper.page.settings.loadImages = false;
        casper.thenOpen("tests/site/index.html", {},function() {
            test.assertEquals(this.resources.length, 1, "one resource found");
            this.wait(200, function() {
                this.capture('page2.png');
            });
        })
    }).then(function(){
        casper.page.settings.loadImages = true;
        casper.thenOpen("tests/site/index.html", {}, function() {
            test.assertNotEquals(this.resources.length, 1, "not only one resource found");
            this.wait(200, function() {
                this.capture('page3.png');
            });
        });
    });

    casper.run(function() {
        test.done();
    });
});

```
you will have 4 captures 2 with the ghost, 2 without any ghost.

hlr1983 commented:
@jefleponot 
It worked when I added `response.setEncoding("binary");`!


[test.tar.gz](https://github.com/casperjs/casperjs/files/853664/test.tar.gz)

jefleponot commented:
@hlr1983 

I success to reproduce the bug with your example
but that is strange that bug appears only onto local css "file:///" resources.

When I try on a website or with http://localhost:5321/....css, no bug appears.
Because css is reloaded ... no more cache

jefleponot commented:
We can redefined page.settings.loadImages 

```js

        Object.defineProperty(page.settings,'loadImages', {
            get : function() { return page.loadImagesValue; },
            set : function(newValue) {
                page.loadImagesValue = newValue;
                if (typeof page.clearMemoryCache !== "undefined") {
                    page.clearMemoryCache();
                }
            }
        });
````

What do you think ?

hlr1983 commented:
@jefleponot 
It's ok.

> We can redefined page.settings.loadImages
> 
>         Object.defineProperty(page.settings,'loadImages', {
>             get : function() { return page.loadImagesValue; },
>             set : function(newValue) {
>                 page.loadImagesValue = newValue;
>                 if (typeof page.clearMemoryCache !== "undefined") {
>                     page.clearMemoryCache();
>                 }
>             }
>         });
> What do you think ?

hlr1983 commented:
Some tests failed:

Without redefined page.settings.loadImages:

`PASS 1204 tests executed in 36.463s, 1204 passed, 0 failed, 0 dubious, 5 skipped.`

With it:

FAIL 1184 tests executed in 38.033s, 1179 passed, 5 failed, 4 dubious, 5 skipped.

Details for the 9 failed tests:

```
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/hooks.js
  onResourceRequested() & onResourceReceived() hook tests
    assert: onResourceRequested() receives image requests
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/viewport.js
  viewport() asynchronous tests
    assertEquals: Casper.viewport() changes width asynchronously
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/visible.js
  visibility tests
    assert: Casper.visible() can detect if an element is visible
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/wait.js
  waitForResource() tests
    fail: Casper.waitForResource() waits for a resource
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/wait.js
  waitForExec() tests
    dubious: waitForExec() tests: 3 tests planned, 2 tests executed
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/casper/xpath.js
  XPath tests
    dubious: XPath tests: 6 tests planned, 7 tests executed
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/tester/assert.js
  Common assertions tests
    assertVisible: Tester.assertVisible() works as expected
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/tester/setup-teardown-async.js
  tearDown() tests
    dubious: tearDown() tests: 1 tests planned, 0 tests executed
In /home/henrique/test5/SISTEMA/casperjs/tests/suites/tester/skip.js
  Skip tests (asynchronous)
    dubious: Skip tests (asynchronous): 1 tests planned, 0 tests executed
```

jefleponot commented:
Hi @hlr1983 
Could you tell me when you think I could review your code :-)

hlr1983 commented:
@jefleponot 
Now it's ready for review! :-)
I'm learning how to use GitHub

jefleponot commented on modules/casper.js:
```diff
@@ -2771,6 +2771,20 @@ function createPage(casper) {
+                page.settings.loadImagesValue = newValue;
+            }
+        });
+        page.settings.loadImages = page.settings.loadImagesValue // Should add this?
```
no : you mustn't redefined page.settings.loadImages

jefleponot commented:
@hlr1983 
commits need to be squash because PR must have only 
- 1 commit : code + tests + docs
or 2 commits : tests , code + docs
or 3 commits : tests, docs, code 

:-) the second is better

please : git rebase -i master
and squash all commits into a single one

hlr1983 commented:
I have found this piece of code when doing the _rebase -i_, and I don't know what it is.

```
        popupPage.onLoadFinished = function onLoadFinished() {
            // SlimerJS needs this line of code because of issue
            // https://github.com/laurentj/slimerjs/issues/48
            // else checkStep turns into an infinite loop
            // after clicking on an <a target="_blank">
            casper.navigationRequested  = false;
            casper.popups.push(popupPage);
            casper.emit('popup.loaded', popupPage);
        };
```

hlr1983 commented:
@jefleponot 

I did  `clearCache()`, and `clearMemoryCache()`. Unfortunately, I have no idea of a way to test it.

regards

jefleponot commented:
@hlr1983 
Everything looks good for merging

Do you plan to add the test_mem.js into tests/suites/caperjs directory ?

hlr1983 commented:
@jefleponot 
Maybe latter I add the test_mem.js. I'm looking for a way of creating a huge image file using js, for loading in page content.

What do you think about suggesting to the engines (phantomJS and slimerJS) to have something like a system.memory variable in the System Module, to show the memory usage?
